source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.cc

```diff
@@ -14,10 +14,13 @@ namespace quic {
 
 QuicMemSliceImpl::QuicMemSliceImpl(QuicUniqueBufferPtr buffer, size_t length)
     : fragment_(std::make_unique<Envoy::Buffer::BufferFragmentImpl>(
-          buffer.release(), length,
-          [](const void* p, size_t, const Envoy::Buffer::BufferFragmentImpl*) {
-            delete[] static_cast<const char*>(p);
+          buffer.get(), length,
+          [allocator = buffer.get_deleter().allocator()](const void* p, size_t,
+                                                         const Envoy::Buffer::BufferFragmentImpl*) {
+            quic::QuicBufferDeleter deleter(allocator);
+            deleter(const_cast<char*>(static_cast<const char*>(p)));
           })) {
+  buffer.release();
```

Comment:
How about use a function to initialize fragment_?

```
std::unique_ptr<Envoy::Buffer::BufferFragmentImpl> ConvertToBufferFragment(
  QuicUniqueBufferPtr buffer, size_t length) {
  const void* p = buffer.get();
  return std::make_unique<Envoy::Buffer::BufferFragmentImpl>(p, length, [buf = std::move(buffer)](
    const void*, size_t, const Envoy::Buffer::BufferFragmentImpl*) {};
}

QuicMemSliceImpl::QuicMemSliceImpl(QuicUniqueBufferPtr buffer, size_t length)
  : fragment_(ConvertToBufferFragment(std::move(buffer), length))
```

You don't need the buffer.release() this way.

You can also change ConvertToBufferFragment to be templated on the buffer type, such that it can be used by the other constructor too(the one accepts a std::unique_ptr<char[]>).

Reply:
I see, it seems BufferFragmentImpl should take the releasor by value and std::move() into its member field. But feel free to do it in the separate change, or just add a TODO.

Reply:
Added TODO for it.